python/tchannel/event.py

```diff
@@ -0,0 +1,92 @@
+# Copyright (c) 2015 Uber Technologies, Inc.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+from __future__ import absolute_import
+from enum import IntEnum
+import logging
+
+log = logging.getLogger('tchannel')
+
+
+class EventType(IntEnum):
+    """Types to represent system events
+
+    Events:
+
+    send_request: before client sends request
+
+    send_response: before server sends response
+
+    receive_request: after server receive request
+
+    receive_response: after server receive response
+
+    """
+    send_request = 0x00,
+    send_response = 0x01,
+    receive_request = 0x02,
+    receive_response = 0x03
+
+
+class EventHook(object):
+    """provide all event hook interfaces
+
+    Customized Hook should should inherit from EventHook class and implement
+    the events' hooks that it wants to listen.
+
+    Example::
+        TraceHook(EventHook):
+            def send_request(self, context):
+                ....
+
+    """
+
+    def send_request(self, context):
+        """event hook for sending request"""
+        pass
+
+    def send_response(self, context):
+        """event hook for sending response"""
```

Comment:
Also, the docs need to contain information on what argument you'll receive. As mentioned before, "context" doesn't mean anything here. Is it the Context as defined in tchannel.context? Is it some arbitrary object. Adding new parameters is cheap. Consider specifying actual separate objects instead of arbitrary parameter objects.